In [1]:
import jax
import jax.numpy as jnp

In [2]:
q = jnp.array([1, 2, 3, 4], dtype=jnp.float32)

For function f

In [3]:
def f(q):
    return jnp.log(jnp.sqrt(jnp.matmul(jnp.transpose(q), q)))

This is consistent with what we calculated for the HW1

In [4]:
df_dq = jax.grad(f)(q)
print(df_dq)

[0.03333333 0.06666666 0.09999999 0.13333333]


For function g

In [5]:
def g(q):
    return q[1:] / (jnp.sqrt(jnp.matmul(jnp.transpose(q[1:]), q[1:])))

These results are conistent with the analytical derivatives!

In [6]:
dg_dq = jax.jacfwd(g)(q)
print(dg_dq)

[[ 0.          0.16008219 -0.03841972 -0.0512263 ]
 [ 0.         -0.03841972  0.12806575 -0.07683945]
 [ 0.         -0.0512263  -0.07683945  0.08324274]]


For function h

In [7]:
def h(q):
    return jnp.arccos(q[0] / jnp.sqrt(jnp.matmul(jnp.transpose(q), q)))

These results are consistent with the analytical derivatives!

In [8]:
dh_dq = jax.grad(h)(q)
print(dh_dq)

[-0.17950548  0.01237969  0.01856953  0.02475937]


For problem 3

In [9]:
import transforms3d
import numpy as np

Part a)

In [10]:
r = np.array([[0.1729, -0.1468, 0.9739],
              [0.9739, 0.1729, -0.1468],
              [-0.1468, 0.9739, 0.1729],])


In [11]:
theta = np.arccos((np.trace(r) - 1) / 2) 
print(theta)

alpha = np.array([[r[2, 1] - r[1, 2]],
                  [r[0, 2] - r[2, 0]],
                  [r[1, 0] - r[0, 1]],]) / (2 * np.sin(theta))
print(alpha)

1.8138318028969072
[[0.57731619]
 [0.57731619]
 [0.57731619]]


The rotation matrix R corresponds to a 1.81 radian rotation around axis of [0.577, 0.577, 0.577]. We can further confirm this by using:

In [12]:
q = transforms3d.quaternions.mat2quat(r)
theta = np.arccos(q[0]) * 2
alpha = q[1:] / np.sin(theta / 2)
print(alpha, theta)

[0.57735027 0.57735027 0.57735027] 1.8138455905719817


Problem 4

If the transition is from A to B:

In [29]:
p = [0, -1, 1, 0]
q = [np.sqrt(3) / 2, 0, 0, -1/2]
q_inv = [np.sqrt(3) / 2, 0, 0, 1/2]
s = [0, 2, 1, 3]



s_a = transforms3d.quaternions.qmult(q, s)
print(s_a)
s_a = transforms3d.quaternions.qmult(s_a, q_inv)
print(s_a)

[ 1.5         2.23205081 -0.1339746   2.59807621]
[ 0.          1.8660254  -1.23205081  3.        ]


Otherwise

In [31]:
p = [0, -1, 1, 0]
q = [np.sqrt(3) / 2, 0, 0, 1/2]
q_inv = [np.sqrt(3) / 2, 0, 0, -1/2]
s = [0, 2, 1, 3]



s_a = transforms3d.quaternions.qmult(q, s)
print(s_a)
s_a = transforms3d.quaternions.qmult(s_a, q_inv)
print(s_a)

[-1.5         1.23205081  1.8660254   2.59807621]
[0.         0.1339746  2.23205081 3.        ]
